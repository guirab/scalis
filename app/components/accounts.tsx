"use client";
import { AccountsContext } from "@/store/context";
import Link from "next/link";
import { useContext, useEffect, useState } from "react";

export default function Accounts() {
  const [showAccounts, setShowAccounts] = useState(false);
  const { accounts, setAccounts } = useContext(AccountsContext);

  useEffect(() => {
    if (!showAccounts) return;
    fetch("/api/accounts")
      .then((res) => res.json())
      .then((data) => setAccounts(data))
      .catch((error) => console.error("Error fetching accounts:", error));
  }, [showAccounts]);

  return (
    <div
      className="flex items-center flex-col w-fit mt-4 group group-[data-show=true] bg-white bg-opacity-30 rounded-lg h-fit p-4"
      data-show={showAccounts}
    >
      <span
        className="text-2xl cursor-pointer w-fit"
        onClick={() => setShowAccounts(!showAccounts)}
      >
        Accounts
      </span>
      <div className="flex flex-col w-full group-data-[show=false]:hidden">
        {accounts.map((account: any) => (
          <div
            key={account.id}
            className="flex w-full mt-4 gap-x-4 justify-between"
          >
            <span>Username: {account.username}</span>
            <Link href="/account/login">View Account</Link>
          </div>
        ))}
      </div>
    </div>
  );
}
