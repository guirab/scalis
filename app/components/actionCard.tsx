import { DepositView } from "./cardView/deposit";
import { WithdrawView } from "./cardView/withdraw";
import { TransferView } from "./cardView/transfer";

export const ActionCard: React.FC<React.HTMLProps<any> & ActionCardType> = ({
  action,
  open,
  setOpen,
  ...props
}) => {
  const currentView = {
    deposit: <DepositView open={open} setOpen={setOpen} />,
    withdraw: <WithdrawView open={open} setOpen={setOpen} />,
    transfer: <TransferView open={open} setOpen={setOpen} />,
  }[action as string];

  return (
    <>
      {open && (
        <div
          className="block w-full mt-4 space-y-4 max-w-[240px]"
          data-testid={`${action}-view`}
          {...props}
        >
          {currentView}
        </div>
      )}
    </>
  );
};
