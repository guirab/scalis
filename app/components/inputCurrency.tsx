export const InputCurrency: React.FC<
  React.HTMLProps<any> & InputCurrencyType
> = ({ value, setValue, ...props }) => {
  const currencyMask = (e: React.ChangeEvent<HTMLInputElement>) => {
    let v = e.target.value;
    v = v.replace(/\D/g, "");
    v = v.replace(/(\d)(\d{2})$/, "$1.$2");
    v = v.replace(/(?=(\d{3})+(\D))\B/g, ",");
    setValue(v);
  };
  return (
    <div className="relative w-full">
      <span className="absolute text-black left-px">$&nbsp;</span>
      <input
        autoComplete="off"
        type="text"
        onChange={(e) => currencyMask(e)}
        value={value}
        name={props.name}
        id={props.id}
        required
        className={`text-black pl-4 ${props.className} w-full`}
      />
    </div>
  );
};
