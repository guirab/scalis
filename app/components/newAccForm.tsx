"use client";
import { useState } from "react";
import { create } from "../actions";

export default function NewAccForm() {
  const [newAcc, setNewAcc] = useState(false);

  async function onClick(e: FormData) {
    await create(e).then(() => setNewAcc(false));
  }

  return (
    <div
      className="flex flex-col items-center w-full mt-4 group group-[data-newacc=true]"
      data-newacc={newAcc}
    >
      <span
        className="text-2xl cursor-pointer w-fit"
        onClick={() => setNewAcc(!newAcc)}
      >
        New Account
      </span>
      <form
        action={onClick}
        className="flex flex-col group-data-[newacc=false]:hidden"
      >
        <label htmlFor="username">Username</label>
        <input
          autoComplete="off"
          type="text"
          name="username"
          id="username"
          required
          className="text-black pl-2"
        />
        <label htmlFor="password">Password</label>
        <input
          autoComplete="off"
          type="password"
          name="password"
          id="password"
          required
          className="text-black pl-2"
        />
        <label htmlFor="checking">Checking Balance</label>
        <input
          autoComplete="off"
          type="number"
          name="checking"
          id="checking"
          required
          className="text-black pl-2"
        />
        <label htmlFor="savings">Savings Balance</label>
        <input
          autoComplete="off"
          type="number"
          name="savings"
          id="savings"
          required
          className="text-black pl-2"
        />
        <button type="submit" className="bg-green-600 mt-4">
          Create Account
        </button>
      </form>
    </div>
  );
}
