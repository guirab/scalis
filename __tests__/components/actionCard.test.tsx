import { render, screen, fireEvent, cleanup } from "@testing-library/react";
import { afterEach, describe, expect, test, vitest } from "vitest";

import { ActionCard } from "../../app/components/actionCard";

describe("ActionCard", () => {
  afterEach(() => {
    cleanup();
  });
  test("snapshot", () => {
    const { container } = render(
      <ActionCard action="deposit" open={true} setOpen={vitest.fn()} />
    );
    expect(container).toMatchSnapshot();
  });

  test("renders DepositView when action is 'deposit'", () => {
    render(<ActionCard action="deposit" open={true} setOpen={vitest.fn()} />);
    expect(screen.getByRole("button").textContent).toBe("Deposit");
  });

  test("renders WithdrawView when action is 'withdraw'", () => {
    render(<ActionCard action="withdraw" open={true} setOpen={vitest.fn()} />);
    expect(screen.getByRole("button").textContent).toBe("Withdraw");
  });

  test("renders TransferView when action is 'transfer'", () => {
    render(<ActionCard action="transfer" open={true} setOpen={vitest.fn()} />);
    expect(screen.getByRole("button").textContent).toBe("Transfer");
  });

  test("does not render any view when open is false", () => {
    render(<ActionCard action="deposit" open={false} setOpen={vitest.fn()} />);
    expect(screen.queryByTestId("deposit-view")).toBeNull();
    expect(screen.queryByTestId("withdraw-view")).toBeNull();
    expect(screen.queryByTestId("transfer-view")).toBeNull();
  });
});
