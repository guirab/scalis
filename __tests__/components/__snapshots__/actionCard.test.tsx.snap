// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActionCard > snapshot 1`] = `
<div>
  <div
    class="block w-full mt-4 space-y-4 max-w-[240px]"
    data-testid="deposit-view"
  >
    <div>
      <label
        for="deposit"
      >
        Into:  
      </label>
      <select
        class="text-black pl-2 outline-none rounded-md w-full"
        id="deposit"
        name="deposit"
      >
        <option
          value="checking"
        >
          Checking
        </option>
        <option
          value="savings"
        >
          Savings
        </option>
      </select>
      <div>
        <label
          for="currency-deposit"
        >
          Amount:  
        </label>
        <div
          class="relative w-full"
        >
          <span
            class="absolute text-black left-px"
          >
            $ 
          </span>
          <input
            autocomplete="off"
            class="text-black pl-4 text-black pl-2 appearance-none outline-none rounded-md w-full"
            data-testid="currency-input"
            id="currency-deposit"
            name="currency-deposit"
            required=""
            type="text"
            value=""
          />
        </div>
      </div>
      <button
        class="capitalize bg-blue-500 text-white px-4 py-2 rounded-md mt-4 w-full"
      >
        deposit
      </button>
    </div>
  </div>
</div>
`;
